File: ecommerce-platform/payment-service/app/messaging.py
# ============================================================
# Payment Service – RabbitMQ Messaging
# Publishes payment events (completed, failed).
# ============================================================
import json
import aio_pika
from app.config import settings

_connection = None
_channel = None
EXCHANGE = "ecommerce_events"


async def connect_rabbitmq():
    global _connection, _channel
    try:
        _connection = await aio_pika.connect_robust(settings.RABBITMQ_URL)
        _channel = await _connection.channel()
        await _channel.declare_exchange(EXCHANGE, aio_pika.ExchangeType.TOPIC, durable=True)
        print("✅ Connected to RabbitMQ")
    except Exception as e:
        print(f"⚠️  RabbitMQ connection failed: {e}")


async def publish_message(routing_key: str, data: dict):
    if not _channel:
        return
    try:
        exchange = await _channel.get_exchange(EXCHANGE)
        message = aio_pika.Message(
            body=json.dumps(data).encode(),
            delivery_mode=aio_pika.DeliveryMode.PERSISTENT,
        )
        await exchange.publish(message, routing_key=routing_key)
    except Exception as e:
        print(f"Failed to publish: {e}")


async def close_rabbitmq():
    if _connection:
        await _connection.close()

File: ecommerce-platform/product-service/src/messaging.js
// ============================================================
// RabbitMQ Messaging for Product Service
// Identical pattern to user-service messaging module.
// ============================================================
const amqp = require('amqplib');

let channel = null;
const EXCHANGE = 'ecommerce_events';

async function connectRabbitMQ() {
  try {
    const url = process.env.RABBITMQ_URL || 'amqp://guest:guest@localhost:5672';
    const connection = await amqp.connect(url);
    channel = await connection.createChannel();
    await channel.assertExchange(EXCHANGE, 'topic', { durable: true });
    console.log('✅ Connected to RabbitMQ');
    connection.on('close', () => setTimeout(connectRabbitMQ, 5000));
  } catch (err) {
    console.error('RabbitMQ connection failed:', err.message);
    setTimeout(connectRabbitMQ, 5000);
  }
}

async function publishMessage(routingKey, data) {
  if (!channel) return;
  channel.publish(EXCHANGE, routingKey, Buffer.from(JSON.stringify(data)), { persistent: true });
}

async function subscribe(routingKey, handler) {
  if (!channel) return;
  const q = await channel.assertQueue('', { exclusive: true });
  await channel.bindQueue(q.queue, EXCHANGE, routingKey);
  channel.consume(q.queue, (msg) => {
    if (msg) { handler(JSON.parse(msg.content.toString())); channel.ack(msg); }
  });
}

module.exports = { connectRabbitMQ, publishMessage, subscribe };
